### DISCUSSION TITLE
Can CRI-O run systemd inside containers?

### DISCUSSION THREAD

**Discussion by christian-2** [cri-o/cri-o#7717] [Q&A]
So Podman can run [systemd in containers](https://blog.while-true-do.io/podman-systemd-in-containers/). Can a Kubernetes cluster that uses CRI-O as its container network do so as well?

**Comment by saschagrunert** [comment]
@christian-2 haven't tried it yet but we have no test case covering that scenario.

May I ask you about the use case of running systemd in Kubernetes? I'm curios which problem you're going to address with that approach.

**Comment by christian-2** [comment]
I am dealing with a HTTP server workload that exists in two versions: one version is running in `apache2.service` (outside containers; with systemd), the other is running in [`docker.io/https`](https://hub.docker.com/_/httpd) (in a Kubernetes cluster; currently with `containerd` as container runtime; without systemd). I'd like to bring the two versions closer together by employing `apache2.service` in both, if feasible. So my expectation is not that the container would become a full "VM", i.e. I would expect it to run `apache2.service` only. Hence the paradigm would not shift from application containers, it's still about a single "process" (i.e. systemd service) inside each container.

Does this make sense? I'd be interested contribute to CRI-O in this area, if applicable.

**Reply by saschagrunert** [reply]
I think I'm lacking a bit of context to judge if it makes sense or not. :upside_down_face: 

You could also go the other way around and run a container using Podman as systemd service. This way you only have to maintain a single image which can be used for Kubernetes as well as the other environment.

I assume that a privileged container is capable of running systemd in CRI-O, but we would have to check the details. An integration test for that use-case would be helpful if we want to support it, though. :+1:

**Reply by christian-2** [reply]
Your suggestion re Podman makes a lot of sense to me, but some (if I may say so) older-minded colleagues still mistrust containers, so I'd like to approach this in steps with the first version (it's a production workload) remaining container-less for the time being. I also agree about a specific integration test.
